lib/keystore.cc

```diff
@@ -82,55 +130,113 @@ rpmRC keystore_fs::delete_key(rpmtxn txn, rpmPubkey key)
+	    delete_key(txn, fp.substr(32), keyfmt);
+	}
     }
 
-    if (rpmMkdirs(rpmtxnRootDir(txn), "%{_keyringpath}")) {
+    free(dir);
+    return rc;
+}
+
+/*****************************************************************************/
+
+static rpmRC acquire_write_lock(rpmtxn txn)
+{
+    char * keyringpath = rpmGenPath(rpmtxnRootDir(txn), "%{_keyringpath}/", NULL);
+    char * lockpath = rpmGenPath(rpmtxnRootDir(txn), "%{_keyringpath}/", "writelock");
+    FILE* fp = NULL;
+    rpmRC rc = RPMRC_FAIL;
+
+    if (rpmMkdirs(NULL, keyringpath)) {
 	rpmlog(RPMLOG_ERR, _("failed to create keyring directory %s: %s\n"),
-		path, strerror(errno));
+	       keyringpath, strerror(errno));
 	goto exit;
     }
 
-    if (FD_t fd = Fopen(tmppath ? tmppath : path, "wx")) {
-	size_t keylen = strlen(keyval);
-	if (Fwrite(keyval, 1, keylen, fd) == keylen)
-	    rc = RPMRC_OK;
-	Fclose(fd);
-    }
-    if (!rc && tmppath && rename(tmppath, path) != 0)
-	rc = RPMRC_FAIL;
+    fp = std::fopen(lockpath, "wx");
 
-    if (rc) {
-	rpmlog(RPMLOG_ERR, _("failed to import key: %s: %s\n"),
-		tmppath ? tmppath : path, strerror(errno));
-	if (tmppath)
-	    unlink(tmppath);
+    if (!fp) {
+	rpmlog(RPMLOG_ERR, _("Can't acquire writelock for keyring at %s\n"), keyringpath);
+    } else {
+	rc = RPMRC_OK;
+	std::fclose(fp);
     }
 
-    if (!rc && replace) {
-	/* find and delete the old pubkey entry */
-	if (delete_key(txn, fp, keyfmt) == RPMRC_NOTFOUND) {
-	    /* make sure an old, short keyid version gets removed */
-	    delete_key(txn, fp+32, keyfmt);
-	}
-    }
+ exit:
+    free(keyringpath);
+    free(lockpath);
+    return rc;
+}
 
-exit:
+static void free_write_lock(rpmtxn txn)
+{
+    char * path = rpmGenPath(rpmtxnRootDir(txn), "%{_keyringpath}/", "writelock");
+    unlink(path);
     free(path);
-    free(keyval);
-    free(keyfmt);
-    free(tmppath);
+}
+
+rpmRC keystore_openpgp_cert_d::load_keys(rpmtxn txn, rpmKeyring keyring)
+{
+    return load_keys_from_glob(txn, keyring, "%{_keyringpath}/*/*");
+}
+
+rpmRC keystore_openpgp_cert_d::delete_key(rpmtxn txn, rpmPubkey key)
+{
+    rpmRC rc = RPMRC_NOTFOUND;
+
+    if (acquire_write_lock(txn) != RPMRC_OK)
+	return RPMRC_FAIL;
+
+    string fp = rpmPubkeyFingerprintAsHex(key);
+    string dir = fp.substr(0, 2);
+    string filename = fp.substr(2);
+    char * filepath = rpmGetPath(rpmtxnRootDir(txn), "%{_keyringpath}/", dir.c_str(), "/", filename.c_str(), NULL);
+    char * dirpath = rpmGetPath(rpmtxnRootDir(txn), "%{_keyringpath}/", dir.c_str(), NULL);
+
+    if (!access(filepath, F_OK))
+	rc = unlink(filepath) ? RPMRC_FAIL : RPMRC_OK;
+    /* delete directory if empty */
+    rmdir(dirpath);
+
+    free(filepath);
+    free(dirpath);
+    free_write_lock(txn);
     return rc;
 }
 
+rpmRC keystore_openpgp_cert_d::import_key(rpmtxn txn, rpmPubkey key, int replace, rpmFlags flags)
+{
+    rpmRC rc = RPMRC_NOTFOUND;
+
+    if ((rc = acquire_write_lock(txn)) == RPMRC_OK) {
```

Comment:
I'd suggest using a "bail out" conditional like in `delete_key()`, it avoids the extra indentation level and would be more consistent with the other method.

Reply:
Yes, I tried that. But with the C++ rules for variable initialization this does not work that nicely here.

Reply:
Nevermind. Using `return` works just fine. Using `goto exit;` doesn't.

Reply:
And changed.